File: doc-examples/r2dbc-example-java/src/main/java/example/BookController.java
package example;

import io.micronaut.http.HttpResponse;
import io.micronaut.http.annotation.*;
import io.reactivex.Single;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;

@Controller("/books")
public class BookController {
    private final BookRepository bookRepository;

    public BookController(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

     // tag::create[]
    @Post("/")
    Single<Book> create(@Valid Book book) {
        return Single.fromPublisher(bookRepository.save(book));
    }
    // end::create[]

    // tag::read[]
    @Get("/")
    Flux<Book> all() {
        return bookRepository.findAll(); // <1>
    }

    @Get("/{id}")
    Mono<Book> show(Long id) {
        return bookRepository.findById(id); // <2>
    }
    // end::read[]

    // tag::update[]
    @Put("/{id}")
    Single<Book> update(@NotNull Long id, @Valid Book book) {
        return Single.fromPublisher(bookRepository.update(book));
    }
    // end::update[]

    // tag::delete[]
    @Delete("/{id}")
    Single<HttpResponse<?>> delete(@NotNull Long id) {
        return Single.fromPublisher(bookRepository.deleteById(id))
                .map(deleted -> deleted > 0 ? HttpResponse.noContent() : HttpResponse.notFound());
    }
    // end::delete[]
}

File: doc-examples/r2dbc-example-python/src/main/python/example/BookController.py
from typing import Annotated

from jakarta.validation import Valid
from jakarta.validation.constraints import NotNull
from micronaut.http import HttpResponse
from micronaut.http.annotation import Controller, Delete, Get, Post, Put
from reactor.core.publisher import Flux, Mono

from example.Book import Book
from example.BookRepository import BookRepository


@Controller("/books")
class BookController:

    def __init__(self, book_repository: BookRepository):
        self.book_repository = book_repository

    # tag::create[]
    @Post("/")
    def create(self, book: Annotated[Book, Valid]) -> Mono[Book]:
        return Mono.from_(self.book_repository.save(book))
    # end::create[]

    # tag::read[]
    @Get("/")
    def all(self) -> Flux[Book]:
        return self.book_repository.findAll()  # <1>

    @Get("/{id}")
    def show(self, id: int) -> Mono[Book]:
        return self.book_repository.findById(id)  # <2>
    # end::read[]

    # tag::update[]
    @Put("/{id}")
    def update(self, id: Annotated[int, NotNull], book: Annotated[Book, Valid]) -> Mono[Book]:
        return Mono.from_(self.book_repository.update(book))
    # end::update[]

    # tag::delete[]
    @Delete("/{id}")
    def delete(self, id: Annotated[int, NotNull]) -> Mono[HttpResponse]:
        return (Mono.from_(self.book_repository.deleteById(id))
                .map(lambda deleted: HttpResponse.noContent() if deleted > 0 else HttpResponse.notFound()))
    # end::delete[]
